aggregator/elem.go

```diff
@@ -0,0 +1,366 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregator
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3aggregator/aggregation"
+	"github.com/m3db/m3metrics/metric"
+	"github.com/m3db/m3metrics/metric/unaggregated"
+	"github.com/m3db/m3metrics/policy"
+)
+
+var (
+	errCounterElemClosed = errors.New("counter element is closed")
+	errTimerElemClosed   = errors.New("timer element is closed")
+	errGaugeElemClosed   = errors.New("gauge element is closed")
+)
+
+type aggMetricFn func(
+	idPrefix []byte,
+	id metric.ID,
+	idSuffix []byte,
+	timestamp time.Time,
+	value float64,
+	policy policy.Policy,
+)
+
+type newMetricElemFn func() metricElem
+
+// metricElem is the common interface for metric elements
+type metricElem interface {
+	// ResetSetData resets the counter and sets data
+	ResetSetData(id metric.ID, policy policy.Policy)
```

Comment:
It not only resets its internal state, but also sets the initial data, hence `ResetSetData`